layers/cmd/ingest_csv.go

```diff
@@ -195,8 +198,268 @@ func (ci *CSVIngester) Run(pipeline *pipeline.PipelineContext) error {
 	return nil
 }
 
+type CSVJoinIngester struct {
+	BaseIngestParams
+	StartRow     int    `json:"startRow" yaml:"startRow"`
+	EndRow       int    `json:"endRow" yaml:"endRow"`
+	HeaderRow    int    `json:"headerRow" yaml:"headerRow"`
+	ID           string `json:"id" yaml:"id"`
+	IngestByRows bool   `json:"ingestByRows" yaml:"ingestByRows"`
+	Delimiter    string `json:"delimiter" yaml:"delimiter"`
+	ColumnRange  uint   `json:"columnRange" yaml:"columnRange"`
+	initialized  bool
+	ingester     *ls.Ingester
+	entities     []CSVJoinConfig
+}
+
+type CSVJoinConfig struct {
+	VariantID        string
+	StartCol, EndCol int
+	IDCols           []int
+}
+
+type joinData struct {
+	CSVJoinConfig
+	data []string
+	id   string
+}
+
+func (cji *CSVJoinIngester) Run(pipeline *pipeline.PipelineContext) error {
+	var layer *ls.Layer
+	var err error
+	if !cji.initialized {
+		layer, err = LoadSchemaFromFileOrRepo(pipeline.Context, cji.CompiledSchema, cji.Repo, cji.Schema, cji.Type, cji.Bundle)
+		if err != nil {
+			return err
+		}
+		pipeline.Properties["layer"] = layer
+		cji.initialized = true
+		cji.ingester = &ls.Ingester{Schema: layer}
+	}
+
+	parser := csvingest.Parser{
+		OnlySchemaAttributes: cji.OnlySchemaAttributes,
+		SchemaNode:           layer.GetSchemaRootNode(),
+		IngestNullValues:     cji.IngestNullValues,
+	}
+	idTemplate := cji.ID
+	if idTemplate == "" {
+		idTemplate = "row_{{.rowIndex}}"
+	}
+	idTmp, err := template.New("id").Parse(idTemplate)
+	if err != nil {
+		return err
+	}
+	if cji.HeaderRow >= cji.StartRow {
+		return errors.New("Header row is ahead of start row")
+	}
+
+	for {
+		entryInfo, stream, err := pipeline.NextInput()
+		if err != nil {
+			return err
+		}
+		if stream == nil {
+			break
+		}
+		pipeline.Context.GetLogger().Debug(map[string]interface{}{"csvingest": "start new stream"})
+		reader := csv.NewReader(stream)
+		if !cji.IngestByRows {
+			pipeline.SetGraph(cmdutil.NewDocumentGraph())
+		}
+		//
+		seenIds := make(map[string][]string)
+		joinCtx := make([]joinData, 0)
+		done := false
+		var doneErr error
+		for row := 0; !done; row++ {
+			func() {
+				defer func() {
+					if err := recover(); err != nil {
+						pipeline.ErrorLogger(pipeline, fmt.Errorf("Error in file: %s, row: %d %v", entryInfo.GetName(), row, err))
```

Comment:
pipeline.ErrorLogger(pipeline, fmt.Errorf("Error in file: %s, %v, %v", xml.Schema, err, string(debug.Stack())))